### PR TITLE
[PT FE]: use example input for dtype and rank detection

### PR DESCRIPTION
**Pull request by eaidova**
### Details:
 - *use example_input for rank and input dtype detection*
 - *support unordered dict inputs*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 1]
```diff
@@ -160,14 +157,14 @@ def prepare_example_inputs(inputs, input_signature):
             if example_inputs is None:
                 scripted = torch.jit.script(pt_module)
             else:
-                inputs, input_signature = prepare_example_inputs(example_inputs, input_signature)
+                input_paremeters, input_signature = prepare_example_inputs(example_inputs, input_signature)
```
Comment (eaidova) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:160]:
```suggestion
                input_parameters, input_signature = prepare_example_inputs(example_inputs, input_signature)
```

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 2]
```diff
@@ -160,14 +157,14 @@ def prepare_example_inputs(inputs, input_signature):
...
                 try:
-                    scripted = torch.jit.trace(pt_module, inputs)
+                    scripted = torch.jit.trace(pt_module, **input_paremeters)
```
Comment (eaidova) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:162]:
```suggestion
                    scripted = torch.jit.trace(pt_module, **input_parameters)
```

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 3]
```diff
@@ -160,14 +157,14 @@ def prepare_example_inputs(inputs, input_signature):
...
                 except Exception:
                     try:
                         scripted = torch.jit.script(pt_module)
                     except Exception:
-                        scripted = torch.jit.trace(pt_module, inputs, strict=False)
+                        scripted = torch.jit.trace(pt_module, **input_paremeters, strict=False)
```
Comment (eaidova) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:167]:
```suggestion
                        scripted = torch.jit.trace(pt_module, **input_parameters, strict=False)
```

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 4]
```diff
@@ -136,21 +136,17 @@ def prepare_example_inputs(inputs, input_signature):
...
                             if key not in inputs:
                                 continue
-                            ordered_inputs.append(inputs[key])
                             used_sign.append(key)
-                        inputs = ordered_inputs
                         input_signature = used_sign
-                    else:
-                        inputs = list(inputs.values())
-                        input_signature = input_signature[:len(inputs)]
+                    return {"example_kwarg_inputs": inputs}, input_signature
```
Comment (slyalin) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:145]: I tried this for dolly with 3 inputs (ids, mask, past_keys_values) and it doesn't work: `jit.trace` complains about too many arguments passed (just ignoring past_key_values input for some reason in the original model signature). The only thing that works is to wrap model.forward to a separate function, define all inputs as positional arguments, and pass it in `example_inputs` instead of `example_kwarg_inputs`. I don't know why it happens, but I cannot reproduce the same issue with your test (where three inputs exist: x, y, z).

Comment (slyalin) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py:145]: Looks like I was able to reproduce: it works only for tensor inputs, and doesn't work for tuples. For some reason, they just cannot be passed in this way. Modified test model (`class NeuralNetwork` defined below) to expect tuple as one of the inputs (z) and pass it in `example_kwarg_inputs` to `jit.trace` and it failed with error: `Exception:
	forward() expected at most 3 argument(s) but received 4 argument(s). Declaration: forward(__torch__.___torch_mangle_1447.NeuralNetwork self, Tensor x, Tensor y) -> Tensor`.

Comment (eaidova) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py:145]: which pytorch version do you use?
I see in optimum code that they uses example_kwarg_input of pkv only if version 2.0.1 and on previous versions it does not work

Comment (slyalin) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py:145]: `torch.__version__` gives `2.0.1+cu117`

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_pytorch.py [hunk 5]
```diff
@@ -595,6 +637,39 @@ def forward(self, x):
...
+    example_input = {"x": torch.zeros((1,3,10,10)), "z": torch.ones((1,3,10,10))}
+    ref_model = make_ref_pt_model_with_optional_inputs([-1, -1, -1, -1], z_exist=True)
+    return net, ref_model, {"example_input": example_input}
+
+
+def create_pytorch_module_with_optional_inputs_case3(tmp_dir):
+    net = make_pt_model_with_optional_input()
+    example_input = {"x": torch.zeros((1,3,10,10)), "z": torch.ones((1,3,10,10))}
+    ref_model = make_ref_pt_model_with_optional_inputs([3, 3, 3, 3], z_exist=True)
+    return net, ref_model, {"example_input": example_input, "input_shape": [[3, 3, 3, 3], [3, 3, 3, 3]]}
```
Comment (rkazants) [inline tests/layer_tests/mo_python_api_tests/test_mo_convert_pytorch.py:658]: Is it a normal case that shapes of examples and `input_shape` to be different? What does it mean and is it described clearly in `help`?

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_pytorch.py [hunk 6]
```diff
@@ -595,6 +637,39 @@ def forward(self, x):
     return net, ref_model, {}
 
 
+def create_pytorch_module_with_optional_inputs_case1(tmp_dir):
+    net = make_pt_model_with_optional_input()
+    example_input = {"x": torch.zeros((1,3,10,10)), "y": torch.ones((1,3,10,10))}
+    ref_model = make_ref_pt_model_with_optional_inputs([-1, -1, -1, -1])
```
Comment (rkazants) [inline tests/layer_tests/mo_python_api_tests/test_mo_convert_pytorch.py:643]: Very interesting that `example_input` does not set input shapes of the generated model.
Is it planned so? May we have confusion from users? I think it is better to describe it in `help`.
Now it says `Sample of model input in original framework.`

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 7]
```diff
@@ -133,24 +134,28 @@ def _get_scripted_model(self, pt_module, example_inputs=None):
...
                             ordered_inputs.append(inputs[key])
                             used_sign.append(key)
-                        inputs = ordered_inputs
                         input_signature = used_sign
                     else:
-                        inputs = list(inputs.values())
-                        input_signature = input_signature[:len(inputs)]
+                        ordered_inputs = list(inputs.values())
+                    if is_torch_2:
+                        return {"example_kwarg_inputs": inputs}, input_signature
```
[the remark this answers is not in the document]
Comment (eaidova) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py:152]: example_kwarg_inputs ([dict](https://docs.python.org/3/library/stdtypes.html#dict), optional) – This parameter is a pack of keyword arguments of example inputs that will be passed to the function while tracing. Default: None. Either this argument or example_inputs should be specified. The dict will be unpacking by the arguments name of the traced function. If the keys of the dict don’t not match with the traced function’s arguments name, a runtime exception will be raised.

it is used in trorch.jit.trace
https://pytorch.org/docs/stable/generated/torch.jit.trace.html

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_pytorch.py (hunk 6 above)
Comment (eaidova) [reply tests/layer_tests/mo_python_api_tests/test_mo_convert_pytorch.py:643]: no, we do not plan restrict input shapes by example_inputs, we use it only for rank prediction. it is needed only for case when reverse infer can not properly setup it for improving user experience

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (eaidova) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py:152]: this dict represents input_parameters in code where we call trace and passed as kwargs to torch.jit.trace. I used such format for minimization difference of call between regular inputs case and kwargs

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_pytorch.py (hunk 6 above)
Comment (rkazants) [reply tests/layer_tests/mo_python_api_tests/test_mo_convert_pytorch.py:643]: I think we need to describe this better in `help` because now we may have some misalignment among FEs.
Just to set rules for all FEs.

@popovaan, do we set `example_input` shapes for model conversion in TF FE or resulted input shapes can be different from `example_input` shapes?

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rkazants) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py:152]: can we use `example_inputs` instead?

[the remark this answers is not in the document]
Comment (eaidova) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py:152]: what do you mean? Just different name of variable or use another parameter for tracing.
If second, no it may lead to incorrect results in case when forward has optional arguments ad provided not all inputs with skipping in middle.
I provided test for this case (when we have input x, y=None, z=None and provided dict with x and z only. using example_nputs will lead that model got for tracing x and y as it accept list of unnamed inputs

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py (hunk 4 above)
Comment (slyalin) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py]: 2.0.1 looks to be the latest official version in PyPI, as you mentioned probably nightly works. OK, let's put tuple issues aside, I will cover it somehow separately.

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py [hunk 8]
```diff
@@ -399,7 +399,7 @@ def convert_params_lists_to_dicts(input_model,
 
         # this cycle adds each unnamed type to dictionary using name from model_inputs
         for idx, node_type in enumerate(input_user_data_types):
-            assert isinstance(node_type, (type, Type)), "Got incorrect format of input types. " \
+            assert isinstance(node_type, (type, np.dtype, Type)), "Got incorrect format of input types. " \
```
Comment (eaidova) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:402]: @rkazants @popovaan please help me to understand, it is bug in the code or wrong error message here?

error says that numpy dtype is expected, but it is not listed in isinstance check, as the result when I use numpy dtype, this function report error. If it is not expected anymore and desired ov.Type, I can switch on it.

Comment (rkazants) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:402]: Great catch! @popovaan, please advice.
